core/tests/protractor_utils/BlogDashboardPage.js

```diff
@@ -0,0 +1,319 @@
+    await action.click('Create Blog Post button', createBlogPostButton);
+
+    await waitFor.pageToFullyLoad();
+    await waitFor.visibilityOf(
+      BlogPostTitleFieldElement,
+      'Blog Post Editor is taking too long to appear.');
+    return await waitFor.pageToFullyLoad();
+  };
+
+  this.setContent = async function(richTextInstructions) {
+    var schemaBasedEditorTag = element(by.tagName('schema-based-editor'));
+    await waitFor.visibilityOf(
+      schemaBasedEditorTag, 'Schema based editor tag not showing up');
+    var richTextEditor = await forms.RichTextEditor(schemaBasedEditorTag);
+    await richTextEditor.clear();
+    await richTextInstructions(richTextEditor);
+    await action.click('Save Content Button', saveBlogPostContentButton);
+    await waitFor.invisibilityOf(
+      saveBlogPostContentButton,
+      'Content editor takes too long to disappear');
+  };
+
+  this.saveBlogPostAsDraft = async function(
+      blogPostTitle, richTextInstructions) {
+    await action.sendKeys(
+      'New blog post title field', BlogPostTitleFieldElement, blogPostTitle);
+    await this.setContent(richTextInstructions);
+    await waitFor.visibilityOf(
+      blogPostContentDisplay, 'Blog Post content not showing up');
+    await action.click('Save as draft Button', saveBlogPostAsDraftButton);
+    await waitFor.visibilityOfSuccessToast('Blog Post Saved Successfully.');
+  };
+
+  this.publishBlogPost = async function(
+      blogPostTitle, richTextInstructions, tags) {
+    await action.sendKeys(
+      'New blog post title field', BlogPostTitleFieldElement, blogPostTitle);
+    await this.selectTags(tags);
+    await this.setContent(richTextInstructions);
+    await waitFor.visibilityOf(
+      blogPostContentDisplay, 'Blog Post content not showing up');
+    await this.setThumbnailImage();
+    await action.click('Publish Blog Post', publishBlogPostButton);
+    await action.click(
+      'Confirm Publish Blog Post button', confirmButton);
+    await waitFor.visibilityOfSuccessToast(
+      'Blog Post Saved and Published Succesfully.');
+  };
+
+  this.selectTags = async function(tags) {
+    for (let i = 0; i < tags.length; i++) {
+      await action.click(
+        'select blog post tag', blogPostTags.get(tags[i]));
+    }
+  };
+
+  this.navigateToPublishTab = async function() {
+    await waitFor.pageToFullyLoad();
+    await waitFor.visibilityOf(
+      publishTab, 'Unable to find publish tab button');
+    await action.click('Publish Tab', publishTab);
+  };
+
+  this.navigateToDraftsTab = async function() {
+    await waitFor.pageToFullyLoad();
+    await waitFor.visibilityOf(
+      draftTab, 'Unable to find drafts tab button');
+    await action.click('Draft Tab', draftTab);
+  };
+
+  this.expectNumberOfDraftsBlogPostsToBe = async function(number) {
+    var isPresent = await blogDashboardIntroMessage.isPresent();
+    if (!isPresent) {
+      await this.waitForDraftsBlogPostsToLoad();
+    }
```

Comment:
It might take some time for the blogDashboardIntroMessage to load. Can you wait for another element that will always be present? Alternatively, you could design your if statement around whether `number > 0`

Reply:
Changed to the div which will contain the message. It will be present if there are no blog posts.

Reply:
I don't think this change fixes the problem. Here's the scenario I'm worried about: Suppose that when this code executes, there are no blog posts in the system (so they will never load), but `blogDashboardIntroMessageContainer` hasn't loaded yet. Since `blogDashboardIntroMessageContainer` hasn't loaded, `isPresent = False` and we enter the if block. Then we wait for blog posts that never load, causing the test to fail.

To fix this, I see two options:

1. Wait for some other element that is present regardless of whether there are any blog posts. This element has to load around the same time as the posts, however.
2. Base the if statement around `number > 0`. In other words, base the if statement around whether we expect there to be blog posts

Reply:
We will actually have blog dashboard intro message to be present only if there are no blog posts created by the user. Before the loading screen goes of we check if the message container is to be present or not.

Reply:
Maybe this schematic will help:

```text

Browser:     +-//--> loading screen disappears -//--> intro message appears -//--> (no blog posts appear) -//-->
                                                |
                                                |
                                                |
Protractor:  +-//--> check blogDashboardIntroMessageContainer.isPresent() -//--> if not present: wait for blog posts -//-->

```

Here I'm illustrating what happens in 2 parallel processes: in the browser and in the protractor tests. The time it takes to advance through each process is non-deterministic, which is why the arrows have `//` in them. Therefore it's possible that the `check blogDashboardIntroMessageContainer.isPresent()` step in the protractor test process comes between the `loading screen disappears` and `intro message appears` steps in the browser process. I use `|` to indicate simultaneity between the two processes.

Note that this is only a problem on really slow systems where it might take a while for the intro message to appear. However, sometimes our CI runners can be really slow! This is especially true now when we are running up against our runners' memory limits. More abstractly, our tests should not rely on something happening really fast in the browser. They should work correctly no matter the speed at which the browser process is running.

Does that help?

Reply:
Yup got it. Thanks for explaining. I have according to whether we expect the intro message to be shown or not changed the tests. I have added a separate function for checking if it is visible. Checking number of blog posts is now independent of it.